### ISSUE TITLE
A way to prevent type inference from happening from multiple points in the code

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#2974]
**Is your feature request related to a problem? Please describe.**
Type inference for a variable or instance variable seems to be determined by all assignments to such variable, which occasionally causes the user to have to hunt down for the one line of code that is accidentally setting the type of that variable to something else. This is probably more annoying in class definitions, where any method could be setting an object field to a different type:

```python3
class MyClass:
    def __init__(self) -> None:
        self.x = 123 # one would expect self.x to be int

    def update_x(self):
        # this accidentally causes self.x's type to be int | float,
        self.x = 1.0
```

**Describe the solution you'd like**
Some ideas:
- A flag similar to `reportConstantRedefinition`, but for variable types. Maybe `reportVariableTypeRedefinition` or a less ambitious `reportInstanceVariableTypeRedefinition`
- Alternatively, a flag to report on instance variable declarations that don't have a type hint

**Comment by erictraut** [comment]
There is already a well-defined way to prevent inference from happening. It involves declaring the intended type with an explicit type declaration.

```python
class MyClass:
    def __init__(self) -> None:
        self.x: int = 123

    def update_x(self):
        self.x = 1.0 # Error
```

**Comment by erictraut** [comment]
I don't know that it would make sense to add a diagnostic check for instance variables that don't have a type declaration. That would be overly strict. The code in your sample above is perfectly type safe, after all.

**Comment by Tomaz-Vieira** [comment]
Is there a reason to allow the type of a variable (specially an instance variable) to have it's type defined over multiple lines of code, though? Even more so if that variable has been defined within `__init__`?

The example above type-checks, but we can concoct examples where this would lead to unexpected runtime behavior (say, maybe you don't expect a decimal point when converting `self.x` to `str`)

I'm fine with adding my own explicit type hints to everything for the sake of safety, but I would prefer not sacrificing the convenience and readability of type inference:

`self.my_value: MyValue = MyValue()`

**Comment by erictraut** [comment]
It's quite typical within Python code for a variable to be assigned multiple times in a class with different values. One very common pattern is to initialize to `None` in the initializer and then later assign to a non-None value in one or more other methods.

**Comment by erictraut** [comment]
This change would be really disruptive for most existing pyright users, so even if it were to be added as an option, that option would need to be off even in strict mode. That means it would rarely get used. You would likely be one of the few people to even know about its existence. We generally don't add features that are so niche.

I'm going to close this request for now. We could consider it again in the future if we receive more signal that other pyright users want such a feature.
